idb/postgres/internal/writer/writer.go

```diff
@@ -142,26 +142,32 @@ func setSpecialAccounts(addresses transactions.SpecialAddresses, batch *pgx.Batc
 //       other things too, so it is not clear we should use it. The only
 //       real benefit is that it would slightly simplify this function by
 //       allowing us to leave out the intra / block parameters.
-func transactionAssetID(txn transactions.SignedTxnWithAD, intra uint, block *bookkeeping.Block) uint64 {
+func transactionAssetID(txn transactions.SignedTxnWithAD, intra uint, block *bookkeeping.Block) (uint64, error) {
 	assetid := uint64(0)
 
 	switch txn.Txn.Type {
 	case protocol.ApplicationCallTx:
-		assetid = uint64(txn.ApplicationID)
+		assetid = uint64(txn.Txn.ApplicationID)
 		if assetid == 0 {
 			assetid = uint64(txn.ApplyData.ApplicationID)
 		}
 		if assetid == 0 {
+			if block == nil {
+				return 0, fmt.Errorf("transactionAssetID(): Missing ApplicationID for transaction: %s", txn.ID())
+			}
 			// pre v30 transactions do not have ApplyData.ConfigAsset or InnerTxns
 			// so txn counter + payset pos calculation is OK
 			assetid = block.TxnCounter - uint64(len(block.Payset)) + uint64(intra) + 1
 		}
 	case protocol.AssetConfigTx:
-		assetid = uint64(txn.ConfigAsset)
+		assetid = uint64(txn.Txn.ConfigAsset)
```

Comment:
Would be nice to add a test that `txn.Txn.ConfigAsset` is actually used to set the `asset` column. The reason why we have this bug is because we didn't have a test like this. :(

Reply:
I added one test that fails until making this change. It doesn't check `ApplicationID`, is there anything else it's missing?